File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  let elementSet = new Set();

  for (let i = 0; i < arr.length; i++) {
    if (elementSet.has(arr[i])) return arr[i];
    elementSet.add(arr[i]);
  }

  return -1;

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));

  console.log("");

  console.log( "Expecting: 5");
  console.log( "=>", findFirstDuplicate([1, 5, 2, 3, 5, 4, 1, 3]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

// Very helpful blog by Flatiron alum at https://dev.to/seanwelshbrown/find-the-first-duplicate-in-a-javascript-array-5da3

// MDN docs for Set object: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # type your code in here
  counter_hash = Hash.new(0)

  arr.each do |el|
    counter_hash[el] = counter_hash[el] += 1

    return el if counter_hash[el] > 1
  end

  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
  puts

  puts "Expecting: 5"
  puts "=>", find_first_duplicate([1, 5, 2, 3, 5, 4, 1, 3])
end

# Please add your pseudocode to this file
# And a written explanation of your solution

# Very helpful walkthrough at https://www.youtube.com/watch?v=ix-cOz8YRGs
